rinja/src/dynamic/child.rs

```diff
@@ -0,0 +1,207 @@
+            exit(1);
+        }
+    };
+    let _ = rt.block_on(async {
+        let stdout = Arc::new(Mutex::new(stdout()));
+        let stdin = Arc::new(Mutex::new(BufReader::new(stdin())));
+        let (done_tx, done_rx) = oneshot::channel();
+        let done = Arc::new(Mutex::new(Some(done_tx)));
+
+        let mut threads = Vec::with_capacity(PROCESSORS);
+        for _ in 0..PROCESSORS {
+            threads.push(spawn(dynamic_processor(
+                Arc::clone(&stdout),
+                Arc::clone(&stdin),
+                Arc::clone(&done),
+            )));
+        }
+
+        done_rx.await.map_err(|err| {
+            std::io::Error::new(ErrorKind::BrokenPipe, format!("lost result channel: {err}"));
+        })
+    });
+    rt.shutdown_timeout(Duration::from_secs(5));
+    exit(0)
+}
+
+async fn dynamic_processor(
+    stdout: Arc<Mutex<Stdout>>,
+    stdin: Arc<Mutex<BufReader<Stdin>>>,
+    done: Arc<Mutex<Option<oneshot::Sender<std::io::Result<()>>>>>,
+) {
+    let done = move |result: Result<(), std::io::Error>| {
+        let done = Arc::clone(&done);
+        async move {
+            let mut lock = done.lock().await;
+            if let Some(done) = lock.take() {
+                let _: Result<_, _> = done.send(result);
+            }
+        }
+    };
+
+    let mut line_buf = String::new();
+    let mut response_buf = String::new();
+    loop {
+        line_buf.clear();
+        match stdin.lock().await.read_line(&mut line_buf).await {
+            Ok(n) if n > 0 => {}
+            result => return done(result.map(|_| ())).await,
+        }
+        let line = line_buf.trim_ascii();
+        if line.is_empty() {
+            continue;
+        }
+
+        let MainRequest { callid, name, data } = match serde_json::from_str(line) {
+            Ok(req) => req,
+            Err(err) => {
+                let err = format!("could not deserialize request: {err}");
+                return done(Err(std::io::Error::new(ErrorKind::InvalidData, err))).await;
+            }
+        };
+        response_buf.clear();
+
+        let mut outcome = Outcome::NotFound;
+        for entry in DYNAMIC_TEMPLATES {
+            if entry.name() == name {
+                outcome = entry.dynamic_render(&mut response_buf, &data);
+                break;
+            }
+        }
+
+        // SAFETY: `serde_json` writes valid UTF-8 data
+        let mut line = unsafe { line_buf.as_mut_vec() };
+
+        line.clear();
+        line.push(b'\n');
+        if let Err(err) = serde_json::to_writer(&mut line, &MainResponse { callid, outcome }) {
+            let err = format!("could not serialize response: {err}");
+            return done(Err(std::io::Error::new(ErrorKind::InvalidData, err))).await;
+        }
+        line.push(b'\n');
+
+        let is_done = {
+            let mut stdout = stdout.lock().await;
+            stdout.write_all(line).await.is_err() || stdout.flush().await.is_err()
```

Comment:
For it to be done, it needs to return an error? Seems a bit strange. At the very least, a code comment would be appreciated here.

Reply:
If the connection to the parent process was interrupted, the child process can exit. Severing the connection is the first step in [killing a subprocess](https://docs.rs/tokio/latest/tokio/process/struct.Child.html#method.start_kill).